packetbeat/protos/mysql/mysql.go

```diff
@@ -603,9 +666,38 @@ func (mysql *mysqlPlugin) receivedMysqlRequest(msg *mysqlMessage) {
 		trans.src, trans.dst = trans.dst, trans.src
 	}
 
+	// try to get query string for Execute statement from cache
+	// and delete statement id for Close statement from cache
+	if msg.statementID != 0 {
+		trans.statementID = msg.statementID
+		stmts := mysql.getStmtsMap(msg.tcpTuple.Hashable())
+		if stmts == nil {
+			logp.Debug("mysqldetailed", "Request execute statement for no stream map. Ignoring.")
+			return
+		}
+		if msg.typ == mysqlCmdStmtExecute {
+			if value, ok := stmts[trans.statementID]; ok {
+				trans.query = value.query
+				// parse parameters
+				parameters := mysql.parseMysqlExecuteStatement(msg.raw, value)
+				trans.param = strings.Join(parameters, "#")
+				logp.Debug("mysqldetailed", "parameters: %s", trans.param)
+			} else {
+				logp.Debug("mysqldetailed", "Request execute statement from unknown prepare statement ID. Ignoring.")
+				mysql.transactions.Delete(tuple.Hashable())
+				return
```

Comment:
As packetbeat can be started at any time, there is quite a chance of packetbeat not knowing about a prepared statement. Plus, the statement cache might already have deleted a still active statement. The prepare command should become a document being published as well. Then we can at least create publish an execute statement (maybe optional) with statement ID + Note warning about the actual query being used is unknown. This information is still helpful as we can still build metrics/visualizations on number of execute statements and actual statement prepare/close (statement churn).

Some application use prepared statements for security purposes, basically creating and closing statements like crazy, all the time. Others might cache statements. It's a great source of information to learn about the applications behavior.

Reply:
It'a big problem because packetbeat can be started at anytime and the mysql server to be audited is on for a long time. Plus, packetbeat should be updated and restarted occasionally.
So there is a quite a  chance of not knowing about a prepared statement.
I also cache the prepared statements in a remote redis server and will read back when restart packetbeat. 
You are right, it's a good choice to publish it instead of ignoring the event.